docs/profiling-workflow-corefx.md

```diff
@@ -389,3 +395,156 @@ PerfView has a built-in support for identifying regressions. To use it you need
 ![Regression Report](img/perfview_24_regression_report.png)
 
 It's recommended to use it instead of trying to eyeball complex Flame Graphs.
+
+### VTune
+
+Intel VTune is a very powerful profiler that allows for low-level profiling:
+
+* provides micro-architecture specific analysis
+* identifies hot spots and bottlenecks
+* utilizes MSRs to get additional low-level hardware information
+
+#### When to use
+
+Let's use PerfView to profile the following app that tries to reproduce [Potential regression: Dictionary of Value Types #25842 ](https://github.com/dotnet/coreclr/issues/25842):
+
+```cs
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+
+namespace ProfilingDocs
+{
+    class Program
+    {
+        const int LastElement = 512;
+
+        static void Main()
+        {
+            Dictionary<int, int> dictionary = Enumerable.Range(start: 0, count: LastElement).ToDictionary(x => x);
+
+            ActualJob(dictionary);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static bool ActualJob(Dictionary<int, int> dictionary)
+        {
+            bool result = false;
+
+            for (int i = 0; i < 3_000_000; i++)
+            {
+                for (int notFound = LastElement + 1; notFound < LastElement * 2; notFound++)
+                {
+                    result ^= dictionary.TryGetValue(notFound, out _);
+                }
+            }
+
+            return result;
+        }
+    }
+}
+```
+
+PerfView tells us that 70% of exclusive CPU time was spent in `FindEntry` method:
+
+![PerfView not is not enough](img/vtune_not_perfview.png)
+
+Once we [open](https://github.com/dotnet/coreclr/blob/4aa65f077778c7c05211825f98684aae9c52caaf/src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs#L363-L460) the source code of `FindEntry` method we can see that:
+
+* there are two execution paths based on whether `IEqualityComparer<TKey>` was provided or not (`comparer == null`)
+* there is a separate execution path for Value Types (`default(TKey)! != null`) and another for reference types
+* JIT can **devirtualize** the calls to `EqualityComparer<TValue>.Default` intrinsic. It means that calls to `Equals` and `GetHashCode` can get **inlined**. Which means that neither VS Profiler or PerfView are going to be able to assign profile samples to source code of the inlined method (the samples are going to be assigned to inlinee - `FindEntry` method).
+
+When we open the Flame Graph we can see that the Call Stack ends at `FindEntry` method. So `Equals` and `GetHashCode` got devirtualized and inlined.
+
+![Inlined](img/vtune_why_not_perfview_flamegraph.png)
+
+**You should start every investigation with VS Profiler or PerfView. When you get to a point where you clearly need information on CPU instruction level and you are using Intel hardware, use VTune.**
+
+#### Identifying Hotspots
+
+Run VTune **as Administrator** and click `New Project`:
```

Comment:
This should call out that Hyper-V should be disabled in the OS and VTx/other virtualization options should be disabled in the BIOS so that instruction based sampling is supported (VTune will generally warn, but sometimes not very visibly).

Reply:
Ah, I see the comment below now